pandas/core/indexes/datetimes.py

```diff
@@ -609,6 +611,8 @@ def _simple_new(cls, values, name=None, freq=None, tz=None,
                            dtype=dtype, **kwargs)
             values = np.array(values, copy=False)
 
+        assert isinstance(values, np.ndarray)
```

Comment:
It was suggested above that we should have an assertion to check whether the input to ``_simple_new`` is actually always an ndarray with the new changes. It turns out that it's still not guaranteed to be an ndarray. In particular, ``_shallow_copy`` sometimes calls ``_simple_new`` with a non-ndarray input. Some of these cases are handled by the code directly above this new assert statement but one case that is not handled is the ``DatetimeIndex`` (i.e., it is not converted to an ndarray). This is why I have put code to convert a DTI in ``_shallow_copy`` to an ndarray, although I realise this may not be the correct way to handle this.

Reply:
can you add .a message on the assert as well
can also ``assert is_integer_dtype(values)``